pkg/boot/survey/gradual.go

```diff
@@ -0,0 +1,110 @@
+package survey
+
+import (
+	"context"
+	"time"
+
+	"github.com/jpillora/backoff"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// GradualSurveyor queries a progressively larger subset of the
+// cluster keyspace.  It is recommended for bootstrapping large
+// clusters where multicast storms may degrade the service.
+//
+// See specs/survey.md for details.
+type GradualSurveyor struct {
+	// Factor controls how long the GradualSurveyor will wait for
+	// peers at each distance. The raw wait time is calculated by
+	// evaluating 'distance * factor', before applying min, max &
+	// jitter.  Defaults to 1.5.
+	Factor float64
+
+	// Min and Max wait durations.  Defaults to 5s and 90s.
+	Min, Max time.Duration
+
+	// If true, use the absolute wait durations calculated via min,
+	// max and factor.  By default, the final wait time is sampled
+	// uniformly from the interval (w/2, w), where 'w' is the wait
+	// duration.
+	DisableJitter bool
+	*Surveyor
+}
+
+func (g GradualSurveyor) FindPeers(ctx context.Context, ns string, opt ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	ctxSurv, cancel := context.WithCancel(ctx)
+
+	found, err := g.Surveyor.FindPeers(ctxSurv, ns, append(opt, WithDistance(uint8(0)))...)
+	if err != nil {
+		cancel()
```

Comment:
I think it is a good practice to call `cancel()` even if the call returns an error, you never know where the context is being used despite the error.

Reply:
Fair point, especially since [`s.c.Send`](https://github.com/wetware/casm/pull/24/files#diff-577b62247a5f574cf760a2c26625599ca3e2adc86bcf6085bbb0f29e55352b8aR415) doesn't cancel the context in the event of an error.

(In fact, I'm going to push a quick fix to handle this case ... just in case.)